pkg/openstack/common.go

```diff
@@ -27,3 +44,301 @@ func EnsureDeleted(ctx context.Context, helper *helper.Helper, obj client.Object
+		if metallbcfg.Endpoint == endpt {
+			return metallbcfg.DeepCopy()
+		}
+	}
+
+	return nil
+}
+
+// ServiceDetails - service details to create service.Service with overrides
+type ServiceDetails struct {
+	ServiceName         string
+	Namespace           string
+	Endpoint            service.Endpoint
+	ExternalEndpoints   []corev1.MetalLBConfig
+	ServiceOverrideSpec []service.OverrideSpec
+	RouteOverrideSpec   *route.OverrideSpec
+	endpointName        string
+	endpointURL         string
+	baseServiceLabels   map[string]string
+	serviceLabels       map[string]string
+	hostname            *string
+	route               *routev1.Route
+}
+
+// GetServiceLabels - returns the labels for the service, used as selector on the route
+func (sd *ServiceDetails) GetServiceLabels() map[string]string {
+	return sd.serviceLabels
+}
+
+// GetEndpointName - returns the name of the endpoint
+func (sd *ServiceDetails) GetEndpointName() string {
+	return sd.endpointName
+}
+
+// GetEndpointURL - returns the URL of the endpoint (proto://hostname/path)
+func (sd *ServiceDetails) GetEndpointURL() string {
+	return sd.endpointURL
+}
+
+// public endpoint
+// 1) Default (service override nil, no definition for public endpt in externalEndpoints)
+// - create route + pass in service selector
+// 2) externalEndpoints != nil for public endpt
+// - pass in override for metallb LoadBalancer service
+// 3) service override provided == LoadBalancer service type
+// - do nothing let the service operator handle it
+// 4) service override provided == ClusterIP service type
+// - create route , merge override with service selector
+//
+// internal endpoint
+// 1) Default (service override nil, no definition for internal endpt in externalEndpoints)
+// - service operator will create default service
+// 2) externalEndpoints != nil for internal endpt, no service override provided
+// - pass in override for metallb LoadBalancer service
+// 3) externalEndpoints != nil for internal endpt, service override provided
+// - merge override with metallb service and pass in override
+// 4) externalEndpoints == nil for internal endpt, service override provided
+// - do nothing let the service operator handle it
+
+// CreateEndpointServiceOverride -
+func (sd *ServiceDetails) CreateEndpointServiceOverride() (*service.Service, error) {
+
+	// get service overrides for the endpoint
+	svcOverride := service.GetOverrideSpecForEndpoint(sd.ServiceOverrideSpec, sd.Endpoint)
+
+	// get MetalLB config data for the endpoint
+	metalLBData := GetExternalEnpointDetailsForEndpoint(sd.ExternalEndpoints, sd.Endpoint)
+
+	// if there is no override and no metallb config for the internal endpoint return
+	if svcOverride == nil && metalLBData == nil && sd.Endpoint == service.EndpointInternal {
+		return nil, nil
+	}
+
+	// set the endpoint name to service name - endpoint type
+	sd.endpointName = sd.ServiceName + "-" + string(sd.Endpoint)
+
+	//  create the service label
+	sd.baseServiceLabels = map[string]string{
```

Comment:
As mentioned elsewhere we need to make sure the same AppSelector is used in the route, service, and deployment/statefulset.

Reply:
it is not really required for functionality. the selector in the service is for service -> pod traffic, which is set by the service operator. this here is just setting the label on the service.

Reply:
thanks you comment about this cleared out how the route connects to the service (by the name of the Service), and how the service connects to the pods (by the selector on the service and the matching label on the Pods)